src/blockdata/block.rs

```diff
@@ -60,6 +60,62 @@ pub struct Block {
     pub txdata: Vec<Transaction>
 }
 
+impl Block {
+    /// check if merkle root of header matches merkle root of the transaction list
+    pub fn check_merkle_root (&self) -> bool {
+        self.header.merkle_root == self.merkle_root()
+    }
+
+    /// check if witness commitment in coinbase is matching the transaction list
+    pub fn check_witness_commitment(&self) -> bool {
+
+        // witness commitment is optional if there are no transactions using SegWit in the block
+        if self.txdata.iter().any(|t| t.input.iter().any(|i| !i.witness.is_empty())) == false {
+            return true;
+        }
+        if self.txdata.len() > 0 {
+            let coinbase = &self.txdata[0];
+            if coinbase.is_coin_base() {
+                // commitment is in the last output that starts with below magic
+                if let Some(pos) = coinbase.output.iter()
+                    .rposition(|o| {
+                        o.script_pubkey.as_bytes().len () >= 38 &&
+                        o.script_pubkey.as_bytes()[0..6] == [0x6a, 0x24, 0xaa, 0x21, 0xa9, 0xed] }) {
```

Comment:
I don't think you need `.as_bytes` for this or the operation above, `Script` already implements indexing and has a `len()` method!

Reply:
changed